chain/chain/src/chain.rs

```diff
@@ -2182,17 +2186,45 @@ impl<'a> ChainUpdate<'a> {
 
         // Check we don't know a block with given height already.
         // If we do - send out double sign challenge and keep going as double signed blocks are valid blocks.
-        if let Ok(other_hash) = self
+        if let Ok(epoch_id_to_hash) = self
             .chain_store_update
             .get_any_block_hash_by_height(header.inner.height)
             .map(Clone::clone)
         {
-            let other_header = self.chain_store_update.get_block_header(&other_hash)?;
-
-            on_challenge(ChallengeBody::BlockDoubleSign(BlockDoubleSign {
-                left_block_header: header.try_to_vec().expect("Failed to serialize"),
-                right_block_header: other_header.try_to_vec().expect("Failed to serialize"),
-            }));
+            // first check if there is already known block of the same height that has the same epoch id
+            if let Some(other_hash) = epoch_id_to_hash.get(&header.inner.epoch_id) {
+                // This should be guaranteed but it doesn't hurt to check again
+                if other_hash != &header.hash {
+                    let other_header = self.chain_store_update.get_block_header(&other_hash)?;
+
+                    on_challenge(ChallengeBody::BlockDoubleSign(BlockDoubleSign {
+                        left_block_header: header.try_to_vec().expect("Failed to serialize"),
+                        right_block_header: other_header.try_to_vec().expect("Failed to serialize"),
+                    }));
+                }
+            } else {
+                for (epoch_id, block_hash) in epoch_id_to_hash.iter() {
```

Comment:
let's remove challenge across different epochs. 
Also finality gadget adds protection to make sure we work of finalized epoch